src/builder/QuarkBuilder.sol

```diff
@@ -51,6 +51,170 @@ contract QuarkBuilder {
 
     /* ===== Main Implementation ===== */
 
+    struct BorrowIntent {
+        uint256 amount;
+        string assetSymbol;
+        uint256 blockTimestamp;
+        address borrower;
+        uint256 chainId;
+        uint256[] collateralAmounts;
+        string[] collateralTokenSymbols;
+        address comet;
+    }
+
+    function borrow(
+        BorrowIntent memory borrowIntent,
+        Accounts.ChainAccounts[] memory chainAccountsList,
+        PaymentInfo.Payment memory payment
+    ) external pure returns (BuilderResult memory /* builderResult */ ) {
+        if (borrowIntent.collateralAmounts.length != borrowIntent.collateralTokenSymbols.length) {
+            revert InvalidInput();
+        }
+
+        uint256 actionIndex = 0;
+        // max actions length = bridge each collateral asset + bridge payment token + perform borrow action
+        Actions.Action[] memory actions = new Actions.Action[](borrowIntent.collateralTokenSymbols.length + 1 + 1);
+        IQuarkWallet.QuarkOperation[] memory quarkOperations =
+            new IQuarkWallet.QuarkOperation[](chainAccountsList.length);
```

Comment:
Non-blocking, but would be good to start using new `ArrayList` data structure for this

Reply:
I will make this update in a follow-up